doc/developer/design/20231213_compute_operator_frontier_tracking.md

```diff
@@ -0,0 +1,187 @@
+To reduce complexity in the Compute code, we choose not to attempt this approach.
+
+Instead, we propose adding a frontier logging operator at the end of each LIR region.
+These operators will observe frontier progress in their inputs and report progress events via loggers connected to the `ComputeState`.
+There is already precedence for such operators, namely the ones that log import frontiers and arrangement sizes.
+
+### Reporting Introspection Data
+
+Existing replica introspection data is exposed through per-replica logging dataflows that export introspection arrangements.
+There are two issues with this approach:
+
+1. In multi-replica clusters, different replicas usually have different data in these arrangements.
+   As a result, introspection arrangements can only be queried by special replica-targeted queries.
+   To collect introspection data from all replicas, clients have to enumerate the existing replicas and issue a separate query for each, yielding bad UX.
+2. When a replica is unhealthy and slow to respond to queries, it will also be slow to respond to introspection queries.
+   This is particularly annoying when one wants to use introspection queries to diagnose why a replica is unhealthy.
+
+Since we want to use operator frontier information to report dataflow hydration progress, and hydration is a period in which a replica is likely slow to respond, issue (2) is particularly problematic.
+
+To avoid these issues, we propose reporting operator frontier information to the compute controller through `ComputeResponse`s instead.
+The compute controller can then write this information into storage-managed collections and expose it to SQL clients that way.
+This solves issue (1) as introspection information from all replicas ends up in the same collection and can be retrieved with a single query.
+Issue (2) is mostly solved because the time logged data spends on the replica is reduced.
+It is still the case that when a replica is completely unresponsive, it likely won't schedule the frontier logging operators, so no new introspection data will make it to the compute controller either.
+However, even in this case introspection data previously written can still be queried.
+
+Note that operator frontier information is not particularly useful on its own.
+To make sense of it, clients also need to know about operators (particularly their names) and the connections between them.
+This information is currently available in per-replica introspection arrangements, but that is not useful if we want to avoid issues (1) and (2) above.
+We therefore propose exposing a subset of the existing per-replica introspection sources through `ComputeResponse`s and the compute controller, rather than through logging dataflows:
+
+* `mz_internal.mz_dataflow_operators_per_worker`
+* `mz_internal.mz_dataflow_channels_per_worker`
+* `mz_internal.mz_dataflow_addresses_per_worker`
+* `mz_internal.mz_compute_exports_per_worker`
```

Comment:
I wonder if this is a good time to workshop this name. What if we just called it `mz_compute_objects`, since it list object IDs for compute dataflows? Or possibly something like `mz_object_dataflows`?

Reply:
We could also repurpose the name `mz_dataflows` for it. `mz_dataflows` is currently a view that has (dataflow_id, name), which doesn't seem all that useful. 

Or maybe `mz_compute_dataflows`, since storage objects are not part of this yet?